collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java

```diff
@@ -212,8 +212,14 @@ private void updateOptionsMenu() {
         // Not ready yet. Menu will be updated automatically once it's been prepared.
         if (optionsMenu == null) return;
 
+        FormController formController = Collect.getInstance().getFormController();
+        boolean isAtBeginning = formController.isCurrentQuestionFirstInForm() && !shouldShowRepeatGroupPicker();
```

Comment:
The logic here has is cleaned up quite a bit after the commit you're commenting on, and the latest code hopefully makes more sense.

In short, you're only at the top/beginning of the hierarchy if you're both <at the first question> and <not showing the repeat picker>. The "up" button should be displayed on every screen of the hierarchy except at the beginning, because then there's nothing to go "up" to.